dropwizard-auth/src/test/java/io/dropwizard/auth/principal/PolymorphicPrincipalEntityTest.java

```diff
@@ -0,0 +1,150 @@
+import javax.ws.rs.container.DynamicFeature;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MediaType;
+import java.io.IOException;
+import java.security.Principal;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.failBecauseExceptionWasNotThrown;
+
+/**
+ * Testing that polymorphic principal entity injection works.
+ */
+public class PolymorphicPrincipalEntityTest extends JerseyTest {
+    private static final String JSON_USERNAME = "good-guy";
+    private static final String NULL_USERNAME = "bad-guy";
+    private static final String JSON_USERNAME_ENCODED_TOKEN = "Z29vZC1ndXk6c2VjcmV0";
+    private static final String NULL_USERNAME_ENCODED_TOKEN = "YmFkLWd1eTpzZWNyZXQ=";
+
+    static {
+        BootstrapLogging.bootstrap();
+    }
+
+    @Override
+    protected DeploymentContext configureDeployment() {
+        forceSet(TestProperties.CONTAINER_PORT, "0");
+        return ServletDeploymentContext
+                .builder(new PolymorphicPrincipalInjectedResourceConfig())
+                .initParam(
+                  ServletProperties.JAXRS_APPLICATION_CLASS,
+                  PolymorphicPrincipalInjectedResourceConfig.class.getName())
+                .build();
+    }
+
+    public static class PolymorphicPrincipalInjectedResourceConfig extends AbstractAuthResourceConfig {
+
+        public PolymorphicPrincipalInjectedResourceConfig() {
+            super();
+            register(PolymorphicPrincipalEntityResource.class);
+            packages("io.dropwizard.jersey.jackson");
+        }
+
+        @Override protected Class<? extends Principal> getPrincipalClass() {
+            throw new AssertionError("getPrincipalClass must not be invoked");
+        }
+
+        @Override protected ContainerRequestFilter getAuthFilter() {
+            return new ContainerRequestFilter() {
+                @Override public void filter(ContainerRequestContext requestContext) throws IOException {
+                    throw new AssertionError("getAuthFilter result must not be invoked");
+                }
+            };
+        }
+
+        @Override protected AbstractBinder getAuthBinder() {
+            final Set<Class<? extends Principal>> principalClassSet = new HashSet<>();
+            principalClassSet.add(JsonPrincipal.class);
+            principalClassSet.add(NullPrincipal.class);
+            return new PolymorphicAuthValueFactoryProvider.Binder<Principal>(principalClassSet);
+        }
+
+        @Override protected DynamicFeature getAuthDynamicFeature(ContainerRequestFilter authFilter) {
+            final Authenticator<BasicCredentials, JsonPrincipal> jsonAuthenticator = credentials -> {
+                if (credentials.getUsername().equals(JSON_USERNAME)) {
+                    return Optional.<JsonPrincipal>of(new JsonPrincipal(credentials.getUsername()));
```

Comment:
Since Java 8 the compiler can infer such explicit type arguments, so `<JsonPrincipal>` is redundant, I beleive.